curl/src/main/scala/org/http4s/curl/WebSocketClient.scala

```diff
@@ -0,0 +1,218 @@
+    @inline def isPing: Boolean = (flags & libcurl_const.CURLWS_PING) != 0
+    @inline def isClose: Boolean = (flags & libcurl_const.CURLWS_CLOSE) != 0
+  }
+
+  def apply(ec: CurlExecutorScheduler, respondToPings: Boolean): Option[WSClient[IO]] =
+    Option.when(CurlRuntime.isWebsocketAvailable) {
+      WSClient(respondToPings) { req =>
+        internal.Utils.newZone
+          .flatMap(implicit zone =>
+            createHandler
+              .evalTap(setup(req, ec))
+          )
+          .map(curl =>
+            new WSConnection[IO] {
+
+              private def send(flags: CInt, data: ByteVector) =
+                internal.Utils.newZone.use { implicit zone =>
+                  IO {
+                    val sent = stackalloc[CSize]()
+                    val buffer = data.toPtr
+                    val size = data.size.toULong
+
+                    throwOnError(
+                      libcurl.curl_easy_ws_send(curl, buffer, size, sent, size, flags.toUInt)
+                    )
+                  }
+                }
+
+              override def send(wsf: WSFrame): IO[Unit] = wsf match {
+                case Close(_, _) =>
+                  val flags = libcurl_const.CURLWS_CLOSE
+                  send(flags, ByteVector.empty)
+                case Ping(data) =>
+                  val flags = libcurl_const.CURLWS_PING
+                  send(flags, data)
+                case Pong(data) =>
+                  val flags = libcurl_const.CURLWS_PONG
+                  send(flags, data)
+                case Text(data, true) =>
+                  val flags = libcurl_const.CURLWS_TEXT
+                  val bv =
+                    ByteVector.encodeUtf8(data).getOrElse(throw InvalidTextFrame)
+                  send(flags, bv)
+                case Binary(data, true) =>
+                  val flags = libcurl_const.CURLWS_BINARY
+                  send(flags, data)
+                case _ =>
+                  // NOTE curl needs to know total amount of fragment size in first send
+                  // and it is not compatible with current websocket interface in http4s
+                  IO.raiseError(PartialFragmentFrame)
+              }
+
+              override def sendMany[G[_]: Foldable, A <: WSFrame](wsfs: G[A]): IO[Unit] =
+                wsfs.traverse_(send)
+
+              override def receive: IO[Option[WSFrame]] = IO {
+                val bufSize = 1024.toULong
+                val buffer = stackalloc[Byte](bufSize)
+                val recv = stackalloc[CSize]()
+                val meta = stackalloc[libcurl.curl_ws_frame]()
+                var payload = ByteVector.empty
+
+                while ({
+                  throwOnError(
+                    libcurl.curl_easy_ws_recv(curl, buffer, bufSize, recv, meta)
+                  )
+
+                  payload = payload ++ ByteVector.fromPtr(buffer, (!recv).toLong)
+                  (!recv) == bufSize
+                }) {}
```

Comment:
Is this blocking?

> * Implemented `WSClient` using the [connect-only approach](https://everything.curl.dev/libcurl/ws/concept) which is more aligned with `WSClient` interface

I'm curious why connect-only approach is a better fit? The callback based approach looks to be non-blocking.

Reply:
Yes, that's a blocking call unfortunately!
Connect-only approach eliminates the need to buffering again on receive side.

Reply:
I'll try callback or event based approach next, as doing it using connect only approach seems to be fruitless.

Reply:
Hmm, is it possible to pause/unpause recv to prevent buffering? This is how the HTTP client is working.
https://github.com/http4s/http4s-curl/blob/9354df6461146ff060f5d029c05b714f71ac7bd6/curl/src/main/scala/org/http4s/curl/CurlClient.scala#L128-L132

Reply:
I'm not sure whether that works with websocket transfer or not as websocket transfers seem to have a different lifecycle. Also websocket docs are very sparse and have some mismatches with published libcurl code too, I was and still am trying to build a working code to start experimenting with options available.

Reply:
I'll try that